lib/avalon/batch/entry.rb

```diff
@@ -176,20 +176,20 @@ def self.offset_valid?( offset )
         true
       end
 
-      def self.attach_datastreams_to_master_file( master_file, filename, captions )
+      def self.attach_datastreams_to_master_file( master_file, filename, datastreams )
         structural_file = "#{filename}.structure.xml"
         if FileLocator.new(structural_file).exist?
           master_file.structuralMetadata.content=FileLocator.new(structural_file).reader
           master_file.structuralMetadata.original_name = structural_file
         end
-        captions.each do |c|
-          next unless c.present? && c[:caption_file].present? && FileLocator.new(c[:caption_file]).exist?
-          filename = c[:caption_file].split('/').last
-          label = c[:caption_label].presence || filename
-          language = c[:caption_language].present? ? caption_language(c[:caption_language]) : Settings.caption_default.language
-          supplemental_file = SupplementalFile.new(label: label, tags: ['caption'], language: language)
-          supplemental_file.file.attach(io: FileLocator.new(c[:caption_file]).reader, filename: filename)
-          supplemental_file.save
+        datastreams.each do |ds|
+          next unless ds.present?
+          supplemental_file = case ds.keys[0].to_s
+                              when /caption.*/
+                                process_datastream(ds, 'caption')
+                              when /transcript.*/
+                                process_datastream(ds, 'transcript')
+                              end
           master_file.supplemental_files += [supplemental_file]
```

Comment:
`supplemental_file` can be nil now that the `next` guard got moved into `process_datastream`.  I think this would probably be easiest to handle here with an inline guard.  Does that make sense to you or is there a better way to handle this?
```suggestion
          master_file.supplemental_files += [supplemental_file] unless supplemental_file.nil?
```

Reply:
Using an inline guard is probably the best option. Other thing that immediately comes to mind would be doing something like `master_file.supplemental_file.compact!` after the loop has finished processing. The guard feels cleaner though.